sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala

```diff
@@ -265,6 +343,47 @@ private[client] class Shim_v0_12 extends Shim with Logging {
     dropIndexMethod.invoke(hive, dbName, tableName, indexName, true: JBoolean)
   }
 
+  // Hive 0.12 doesn't support permanent functions. Here we use a in-memory map to simulate the
+  // permanent function registry.
+  private var functionMap = Map.empty[String, CatalogFunction]
+  private val functionDDLNotSupported = "Hive 0.12 doesn't support creating permanent functions. " +
+    "Please use Hive 0.13 or higher."
```

Comment:
For Hive 0.12, let's throw an exception if users want to create a permanent function.